Add basic search functionality, and a small UI bound to mod+f.

stephencookdev opened the pull request:
The search functionality supports single-line, case-sensitive, directional search - with single highlighted results at a time.

This should address https://github.com/zeit/hyper/issues/140 and https://github.com/zeit/hyper/issues/541

This is ready to merge, in theory - but I would be surprised if there are no comments surrounding:

- my use of hterm (never used the library before, I've rewritten some private methods to give us the functionality we need, but hopefully I'm missing something more elegant)
- my styling (I was unsure how to incorporate themes into this, so just tried to aim for a UI that would fit any theme)

I've also made this as minimal a solution as possible, definitely after this I'd really want to expand the feature to support case-non-sensitive / regex options, and also to support "(n results found)" and multi-result highlighting.

For now, though - here's what it looks like. Please check out the branch and try to break it :)

![results](https://cloud.githubusercontent.com/assets/8496655/25780979/660c7ba8-332a-11e7-9e52-4b79d5de7650.gif)

![errors](https://cloud.githubusercontent.com/assets/8496655/25780981/6b96df6e-332a-11e7-8167-381573b6a8fd.gif)

![screenshot from 2017-05-07 10-50-48](https://cloud.githubusercontent.com/assets/8496655/25780987/831a761e-332a-11e7-94cf-ad53c77b6354.png)

![screenshot from 2017-05-07 10-50-59](https://cloud.githubusercontent.com/assets/8496655/25780988/857008fc-332a-11e7-8240-287cab4a7d0e.png)

![screenshot from 2017-05-07 10-52-29](https://cloud.githubusercontent.com/assets/8496655/25780990/88a8aeb6-332a-11e7-8671-78b2414d59e5.png)

![screenshot from 2017-05-07 10-52-38](https://cloud.githubusercontent.com/assets/8496655/25780991/8f0c006e-332a-11e7-9755-052fddadfc09.png)

Features:

- Mod+F to open find
- Mod+F/Esc to close find
- Click "Find" or push "Enter" to find (cycling through 1 at a time) matching search
- Push Shift+Enter to cycle through in reverse
- Searching is per terminal, so only the focused tab/pane/window is actually searched

ppot commented:
When you are searching. You need to select every result at the same time. and using `enter` key to cycle between them. `cursor select the last one` then move to bottom and go back to previous find from each enter key.

ppot commented:
For the hterm lib part. Move you're code to utils/search.js 

export functionality and set them to hterm after. Something like:

`hterm.Terminal.prototype.find  = search.find(searchTerm, options = {})`

stephencookdev commented:
@ppot so currently yes, it's single-select, but pushing enter cycles through the results. Pushing shift+enter cycles through the results in reverse.

And panes are a good point, I should have mentioned. The UI looks exactly the same, and just only the active panes are searched (the same is true of tabs and windows). The search bar is positioned in the top right of the window, not the top right of the pane. I guess this could be confusing, but in my mind it's the window that has a search bar, not the pane.

I'll update the PR description with that^ info.

As for moving code to `utils/search.js` - what's the reasoning there? It seems unnecessary to declare a utility function, and then bind it to the `Terminal` prototype anyway. Any time anyone wanted to use the utility function `find(terminal, str)` they could also just do `terminal.find(str)`. So the API is exactly the same on the `Terminal` prototype, just with more duplication

ppot commented:
For moving those function to utils/search.js is to reduce the complexity of hterm.js file by moving specific parts to their file. But it's okay if it get to big it can be moved afterwards.

ppot commented on app/accelerators.js:
```diff
@@ -24,6 +24,7 @@ const applicationMenu = { // app/menu.js
   paste: 'V',
   selectAll: 'A',
   clear: 'K',
+  find: 'F',
```
you have to `find: 'F'`

stephencookdev commented:
Okay @ppot - I've moved that logic out to utils/ in d6c44cec67e258e54da86fc7284f9e09b29a5021 :)

stephencookdev commented:
Any more comments/reviews on this...?

ppot commented:
Been busy sorry, @albinekb ?

MartyGentillon commented on lib/utils/search.js:
```diff
@@ -0,0 +1,55 @@
+    return true;
+  }
+  return false;
+}
```
50 line functions are a clear code smell.  This needs refactoring.

MartyGentillon commented on lib/utils/search.js:
```diff
@@ -0,0 +1,55 @@
+  let anchorRowNode = sel.anchorNode;
+  while (anchorRowNode && !('rowIndex' in anchorRowNode)) {
+    anchorRowNode = anchorRowNode.parentNode;
+  }
```
Duplicate code

MartyGentillon commented on lib/components/searchbox.js:
```diff
@@ -0,0 +1,190 @@
+    if (term) {
+      const searchTerm = this.state.searchTerm;
+      const options = {
+        dir: this.state.dir
```
dir is a poor name, do you mean directory or direction?  (I am going to guess directory every time, especially when working with a console application).  Avoid abbreviations, it's more important that your code be easily readable than easily typable.

MartyGentillon commented on lib/utils/search.js:
```diff
@@ -0,0 +1,55 @@
+
+  let foundIndex;
+  let curRow;
+  for (let i = 0; i <= rowCount; i++) {
```
Extract this for loop to it's own function.

MartyGentillon commented on lib/utils/search.js:
```diff
@@ -0,0 +1,55 @@
+
+    return true;
+  }
+  return false;
```
The basic find function should probably just return the detailed results of the search.  That will allow it to be composed into other useful functions.  Leave the scrolling and highlighting to another function.

MartyGentillon commented on lib/components/searchbox.js:
```diff
@@ -0,0 +1,190 @@
+        dir: this.state.dir
+      };
+      const searchResult = term.find(searchTerm, options);
+      this.setState({searchError: !searchResult});
```
How are you handling wraparound for the search?

stephencookdev replied:
Do you mean terminal wraparound (not line wrap-around?)

Currently the `find` method will always find the next item, and wraparound implicitly with no feedback to the user.

Ideally there should be a notification of this wraparound having happened to the user, but I want to address this in a separate PR

MartyGentillon replied:
Cool, glad to hear you are thinking about it.

MartyGentillon commented on lib/components/searchbox.js:
```diff
@@ -0,0 +1,190 @@
+
+      searchInput: {
+        fontSize: '16px',
+        padding: '10px',
```
This should scale with the user interface.

MartyGentillon commented on lib/components/searchbox.js:
```diff
@@ -0,0 +1,190 @@
+        padding: '0 15px',
+        background: 'transparent',
+        color: '#555',
+        fontSize: '16px',
```
This should scale with the user interface.

MartyGentillon commented on lib/utils/selection.js:
```diff
@@ -50,4 +52,72 @@ exports.all = function (terminal) {
+}
+
+/*
+ * Copy-pasted from hterm's hterm.Screen.prototype.getNodeAndOffsetWithinRow_
```
I have to wonder if, rather than copying and pasting the code (duplicate code is a great way to make sure that a bug doesn't really get fixed), could you just use the function declared there?

something like
```
var GetNodeAndOffsetWithOverflow = hterm.Screen.prototype.getNodeAndOffsetWithOverflow_
```

stephencookdev replied:
I could, and I originally had something similar when developing

But I really don't like the idea of adding an explicit dependency to private methods in hterm. The maintainers of hterm would be well within their rights to remove/rename the private method, and mark it as only a minor change

MartyGentillon replied:
Yep, this is a situation where there isn't really any good answer....  On one hand, you would sort of like to pick up their bug fixes, on the other hand, it is always best to avoid relying on internals.  I suppose this is probably the better compromise, though I still stubbornly wonder if there isn't a way to do both.  (Perhaps we should file a feature request on them?)

MartyGentillon replied:
Here is a thought for a different compromise, use the function declared there and write a test to make sure that it is there.  Now, if they remove/rename the function, we will know about it, while if they fix a bug, we will get it.

stephencookdev replied:
I get what you're saying, and I can see the benefit - it just feels really weird to me to do this.

I would much rather go the route of taking the (fairly small) duplication hit for now, and lobby hterm to make those methods slightly more public

MartyGentillon replied:
Either or, for the most part, this is about considering options.  Whichever you implement will work fine.

MartyGentillon commented on lib/components/searchbox.js:
```diff
@@ -0,0 +1,190 @@
+        position: 'absolute',
+        top: '50px',
+        right: '50px',
+        width: '400px',
```
This should probably scale with the user interface.

stephencookdev replied:
Yeah, you're right. Do you know of any examples in the codebase where logic like this happens? If not off the top of your head, no worries - I'll dig around

MartyGentillon replied:
I would think that using em would cause it to happen magically.

oalhait commented:
Any updates on this/idea on when it'll get pushed to users?

stephencookdev commented:
I should have time to finish addressing comments + resolving merge conflicts tonight

ppot commented:
Before fixing thing. You should rebase with master since lots of thing changed.

stephencookdev commented:
@ppot yeah, rebased - everything looks like it's working, after a bit of a fix up

@MartyGentillon should have addressed all of your comments.

UTs still to come, I didn't realise quite how bare the UTs were! :(

MartyGentillon commented on lib/utils/search.js:
```diff
@@ -0,0 +1,58 @@
+  };
+}
+
+// find a searchTerm, highlighting it (and scrolling to it if currently off screen)
```
This comment needs to be updated now (you moved the scrolling away from this function.)

stephencookdev commented:
Okay @MartyGentillon I addressed that last comment

I looked again at adding UTs, but I really think it's gonna make more sense to address this in a separate PR, specifically looking at building up the UT framework here

It looks like the current ava set-up isn't even set up to handle the ES6+ syntax used in the project :/ A lot of the decisions

MartyGentillon commented:
Looks good to me.

chabou commented:
It works really great !
Awesome work @stephencookdev!

stephencookdev commented:
@chabou awesome, thank you :)

Is there anything else we need to do to get this merged?

chabou commented:
Just for the record, a plugin exists: https://www.npmjs.com/package/hyper-search

ItamarGronich commented:
Hey, guys? what's the preventing this from merging? i don't want to install a plugin that will later conflict with the build when this gets merged.

jq-87 commented:
The previously mentioned plugin seems to have issues - it does not work for me and other people are also saying the same thing.

They have an open issue:
https://github.com/jaanauati/hyper-search/issues/12

So hopefully this feature gets merged in soon 🙏

stephencookdev commented:
Fixed the merge conflicts

Just a note that this PR has been open for 5 months now, though. I'm not sure who the best people to poke are to look at this, but if anyone knows - please poke away... :)

stshan commented:
Any updates on this? It seems kinda weird that such an important feature PR would be ignored for so long.

pratik0809 commented:
+1 would like to see this feature merged in

david-a commented:
+1 for this one! please merge !

stshan commented:
ping

chabou commented:
We do not use `hterm` and DOM anymore, but `xterm` and `canvas`.

This PR will not work as it with canary, `hyper-search` plugin neither.

We maybe should use xterm search addon. https://github.com/xtermjs/xterm.js/tree/master/src/addons/search

@stephencookdev I am very sorry to close this PR 😞  And thank you so much for your help 🙏

rightaway commented:
So there's really no way to search at all in canary, using a plugin or not?

albinekb commented:
Not yet @rightaway, but as chabou stated it is possible to do this a lot easier now with xterm & xterm search addon.

RichardBronosky commented:
"Possible" in January. But now it's September.

albinekb commented:
Yes, PR welcome 👍  @RichardBronosky We are all busy.

karlhorky commented:
Looks like #3075 ended up fixing this and also #541 .
